Support simple Django forms

sdementen opened the pull request:
hello @trco , could you look at this WIP ?

Some question I have is about the naming of the views. I have added the simplest View as
```
class BSModalView(PassRequestMixin, generic.FormView):
    pass
```
which names look OK to me.

The simples Form is then
```
class BSForm(PopRequestMixin, forms.Form):
    pass
```
but this breaks the prepending of all objects by BSModal.
However the `BSModalForm` is already used for what should have been probably named `BSModalModelForm`.
Is it worth to break the backward compatibility and do this renaming ? or do you prefer to keep the slightly inconsistent (and a bit puzzling to me when I was trying to make sense of the package) ?
Personnally, I think it would be better to rename for the LT.
But this may require a bump of the major version number of the lib.

sdementen commented:
I have also added the tests. Could you check the tests as on my machine (windows 10) I need to use chromedriver and have some issues of reliability.

sdementen commented:
I will finish adapting the README.rst once the renaming point is clearer as the structure may need to change a bit.

trco commented:
@sdementen Big thanks for your efforts! You made my day with your enthusiasm. `BSModalView` sounds perfect to me and I also agree to update current `BSModalForm` to `BSModalModelForm` and use `BSModalForm` for your new implementation. We will bump to 2.0.0, that's not a problem. I'm also working on improvements on #74 and I have reorganized and renamed some things in .js file along the way. This will be also in a new release.

I have checked your example and got an idea of implementing a filter form instead of comment form. Somehow the comment form seems unrelated to the examples. Filter form I'm thinking of would for example have just one dropdown where you would select book type and after redirection the table below would filter only the books of this type. In case there would be no books of selected type, the form could either return a non field error in modal or redirect to the same page with warning message that there is no books of this type in the table. I would be really glad if you could change this. You /simple url could then be /filter, your form a BookFilterForm etc.

Two comments on style, please use single instead of double quotes in .py files and limit the line width to 79 characters.

Can't wait to include this work and have a first human collaborator on this package. Dependabot doesn't count of course ;)

Great work!!!

sdementen commented:
Great, I'll do that tomorrow morning

sdementen commented:
@trco , could you explain why there are differences in the inherited classes of these 3 views
```
class BSModalUpdateView(PassRequestMixin, SuccessMessageMixin, generic.UpdateView):
    pass


class BSModalReadView(generic.DetailView):
    pass


class BSModalDeleteView(DeleteMessageMixin, generic.DeleteView):
    pass

```

sdementen commented:
@trco  regarding single vs double quotes, I am using black as reformatter and it uses double quotes.
is it a strong requirements or could we "blackify" everything ? this is practical as there is no need anymore to clean up the files manually.
but if you prefer other formatting standards, no issue to fix it manually to single quote

sdementen commented:
@trco regarding the filter form (vs the simpleform), I agree that it would fit better the 'demo' (in examples/). Yet, when I was understanding how the package was working, I was more interested in the simplest unit example. In this case, just popping a ModalForm with some useless form would pass the best the message.
If you prefer to have a coherent demo (over a simplest demo), tell me and I will adapt according to your filter scenario

sdementen commented:
Also, regarding the javascript, I think it would be useful to tell the user he/she can add something like the following to his/her page
```javascript

    <button class="bs-modal-popup btn"
                  data-bs-modal-url="{% url 'name-of-bsmodal-view' %}"
                  type="button" 
                  name="button">
          My button to trigger the BSModalView
     </button>

    <script type="text/javascript">
        $(document).ready(function () {
            // activate modal popups
            $(".bs-modal-popup").each(function () {
                $(this).modalForm({
                    formURL: $(this).data('bs-modal-url'),
                    errorClass: ".has-error"
                });
            });
        });
    </script>
```
to illustrate:
- generic assignment of modalForm based on a specific class (bs-modal-popup here)
- use of data-bs-modal-url to bind to the formUrl
- customise the errorClass to .has-error (at least for me with crispy and bootstrap3)

trco commented:
# Difference in generic views
BSModalCreateView and BSModalUpdateView both process the form, BSModalReadView just shows data in modal, while BSModalDeleteView deletes object via post method. They all inherit only what they need to work.

# Single vs. double quotes
I prefer single quotes in .py files so please don't change that. Black is just one of formatters.

# Filter vs. comment form in examples
You've said it all with "...useless form...". Why would someone need a useless form? ;) Make a useful form in example. It doesn't have to bi filter form. It was just proposition.

# Regarding the javascript
It's just another way of writing "How to use" but interested user still has to read about each field. I don't see a benefit in this.

sdementen commented:
(answers in same order)

Thanks for the explanation, after reading the classes, it makes sense. What I did not notice at first is that you are not obliged at all to use any of the generic to use BSModal. The magic is in the javascript side, correct ?

There are " in other .py files (I haven't touched), I changed them to '. I hope it is OK.

For the form, I have added a new filter form, again the simplest possible, the objective being to illustrate the use of the package, not to build a real webapp. Could you tell me if it looks OK ?
I have lot of troubles to make the tests run with selenium and the Chrome driver (I can't have firefox installed on my laptop) so the tests are not yet ok.

Re the javascript, I have slightly adapted the index.html to use the 'each' example on a class bs-modal. Btw, I wonder if the classes "update-book","create-book", etc should not be IDs instead of classes in the html.

sdementen commented:
any feedback @trco  ? tx!

trco commented:
@sdementen Sorry for being unresponsive. Quite bussy at the moment. I'll check everything during the weekend. Tnx for your efforts!!!

trco commented:
@sdementen I have been checking your PR but I think this is not the latest code. Did you maybe forgot to push your latest changes?

sdementen commented:
sorry, indeed, forgot to push. Do you see them now ?

trco commented:
@sdementen Yes. Thanks. I'll check it out in the afternoon.

trco commented on examples/views.py:
```diff
@@ -16,18 +24,51 @@ class Index(generic.ListView):
+
+        # call the base form_valid (that will call the get_success_url)
+        resp = super().form_valid(form)
+        return resp
```
Please rename to `response`. I'm quite obsessed with readability of the code ;)

sdementen replied:
done

trco commented on examples/forms.py:
```diff
@@ -2,13 +2,26 @@
 
 
-class BookForm(BSModalForm):
+class SimpleModalForm(BSModalForm):
```
Please remove SimpleForm at all places. I think that FilterForm you implemented is awesome.

sdementen replied:
done

trco commented:
@sdementen I checked your PR. Awesome work! I made some minor comments. After you resolve this comments and fix the functional tests I'll gladly merge your work.

As you already said functional tests are broken due to chromedriver. Errors give a helpful insight. You just have to find a way to interact with the elements.

However I suggest to revert back to Firefox for now. I have reverted and all the tests pass.

sdementen commented on CHANGELOG.rst:
```diff
@@ -2,6 +2,13 @@
+=====
+
+- add generic view BSModalFormView
+- renamed BSModalForm to BSModalModelForm
```
done

sdementen commented on README.rst:
```diff
@@ -254,9 +254,9 @@ BSModalDeleteView
-BSModalForm
+BSModalModelForm
     Inherits PopRequestMixin, CreateUpdateAjaxMixin and Django's forms.ModelForm.
 
```
done

sdementen commented on CHANGELOG.rst:
```diff
@@ -38,7 +45,7 @@ Changelog
 
 - add generic views BSModalCreateView, BSModalUpdateView, BSModalReadView, BSModalDeleteView, BSModalLoginView
-- add form BSModalForm
+- add form BSModalModelForm
```
yes, sorry, due to a replace all kind of refactoring

sdementen commented on examples/views.py:
```diff
@@ -1,13 +1,21 @@
+    BSModalUpdateView,
+    BSModalReadView,
+    BSModalDeleteView,
+    BSModalFormView,
```
it is simpler to add new imports in this layout as you can just add a line ending with a comma.
otherwise (see for instance https://stackoverflow.com/questions/11597901/why-are-trailing-commas-allowed-in-a-list )
but if you want to stick to the other layout, OK.
btw, in the `urlpatterns`, there is also a trailing comma (which is very fine IMHO)

sdementen commented on examples/views.py:
```diff
@@ -16,18 +24,51 @@ class Index(generic.ListView):
+            self.filter = ''
+        else:
+            # the user has filtered the list of books
+            self.filter = f'?type={form.cleaned_data["type"]}'
```
done

sdementen commented:
@trco  could you have a new look at the PR ?
I cannot run the tests reliably on my environment. Could you confirm they work ?

trco commented on examples/templates/index.html:
```diff
@@ -110,27 +114,27 @@ <h4>
 
-      // Create book button, uses the <div> with id="create-modal"
-      $(".create-book").modalForm({formURL: "{% url 'create_book' %}", modalID: "#create-modal"});
+          // Create buttons with the bs-modal class, uses the <div> with id="modal" (default)
```
Here you should change "Create buttons" to "Update, Read and Delete button".

trco commented on examples/templates/index.html:
```diff
@@ -110,27 +114,27 @@ <h4>
+      $(function () {
+          // Log in & Sign up buttons
+          // The formURL is given explicitly
+          $(".simple-btn").modalForm({formURL: "{% url 'simple' %}"});
```
Simple form should be removed. This breaks index page, since the simple form url was removed. The simple.html is still in templates and on index page. I guess the instantiation of filter form is now also missing in index page js script.

sdementen commented:
I hope I haven't forgotten again sthing ...

trco commented:
@sdementen Thank you for all your effort and for being first contributor.

sdementen commented:
You're welcome!
